File: 0784. Letter Case Permutation.java
// Medium
// Backtracking
// Time:O(n*2^m),Space:O(n)
// https://leetcode.cn/problems/letter-case-permutation/

import java.util.*;

class Solution {
    public List<String> letterCasePermutation(String s) {
        List<String> res = new ArrayList<>();
        backtracking(s.toCharArray(), 0, res, new StringBuilder());
        return res;
    }

    private void backtracking(char[] chars, int index, List<String> res, StringBuilder cur) {
        if (index == chars.length) {
            res.add(cur.toString());
            return;
        }

        char c = chars[index];
        // for each char, it can be lowercase or uppercase
        if (Character.isLetter(c)) {
            // op1: lowercase
            cur.append(Character.toLowerCase(c));
            backtracking(chars, index + 1, res, cur);
            cur.deleteCharAt(cur.length() - 1);

            // op2: uppercase
            cur.append(Character.toUpperCase(c));
            backtracking(chars, index + 1, res, cur);
            cur.deleteCharAt(cur.length() - 1);
            
        } else {
            // if digit, add directly
            cur.append(c);
            backtracking(chars, index + 1, res, cur);
            cur.deleteCharAt(cur.length() - 1);
        }
    }
}

File: 0501-1000/0784. Letter Case Permutation.py
# Time:O(n*2^m), Space:O(n)

from typing import List

class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        res = []

        def backtracking(index, path):
            if index == len(s):
                res.append(path)
                return

            char = s[index]
            if char.isalpha():
                backtracking(index + 1, path + char.lower())
                backtracking(index + 1, path + char.upper())
            else:
                backtracking(index + 1, path + char)

        backtracking(0, "")
        return res
    